core/domain/stats_jobs_one_off.py

```diff
@@ -303,6 +315,25 @@ def reduce(exp_id, values):
 
     @classmethod
     def prepare_reduce(cls, exp_id, values):
+        """Runs the reduce step to extract aggregate exploration statistics from
+        the dicts passed in by the mapping function.
+
+        Args:
+            exp_id: str. The exploration id.
+            values: list(dict(str, str)). List of dicts of events for preparing
```

Comment:
> Done. I'm sure about its type. Please see the for loop just below it. The values is a list var containing value of type dict evident by how we're accessing it by keys- version, event_type etc.

But isn't there an ast.literal_eval() applied before that? What does that call do?

Reply:
`ast.literal_eval` just raises an exception if the input isn't a valid Python datatype, so the code won't be executed if it's not. Please see [here](https://stackoverflow.com/questions/15197673/using-pythons-eval-vs-ast-literal-eval). So it has nothing to do with `list(str)`. Also, in the following for loop, it's clear that it's a list of dicts.

Reply:
ast.literal_eval acts on values and stores it (confusingly) in another variable called values. But it does have an effect.

Try it in your Python console. Open the interpreter and type


```
import ast
ast.literal_eval('2')
```

The second line will produce a value of type int (check and see). But the input is a str.

So ast.literal_eval() changes the type of things. So you need to do a bit more investigation to see what the type is, probably by running the job yourself (from the admin console) with at least one stats event already created, and logging.

Reply:
Yeah, I tried and got the same response. Please look [here](https://docs.python.org/2/library/ast.html#ast.literal_eval). `ast.literal_eval()` safely evaluates the node or string irrespective of the type. So, it should be `list(dict)` as it was previously.

Reply:
I checked after printing its type and it's coming to be `list(str)`. Please follow the following steps to repro:
* Add `raise Exception(type(values))` to print the type of `values` arg.
* Try to create some stats event. I created a new lesson and played it in learner dashboard.
* After that, go to 'Admin' page to run the one off job.
* From the list, select `RecomputeStatisticsOneOffJob` job and run it.
* Check the terminal for the Exception.